### PR TITLE
Move docker image management and test entrypoint to Maven

### LINKED ISSUE #30: Don't build docker images in two places
Right now we have docker image building code in both Scala when run from Maven, and the runner script as it's just invoked from the command line. We should only have building and pushing logic in either place.

My proposal would be that the tests need to be manually passed a pre-built image, and the Maven pre-integration-test phase handles building the images with the right tag and pushing to the right location. Building should only be done in a shell script.

### PR DESCRIPTION
**Pull request by mccheah**
Closes #30.

### CODE REVIEW

**Comment by mccheah** [comment]
Requires both #29 and #14.

**Comment by mccheah** [comment]
Also requires the test commands in Jenkins to change.

**Comment by mccheah** [comment]
todo and not handled by this patch:

- Clean up docker images in the `post-integration-test` phase
- Use your own docker images (it's theoretically possible but the usage of the examples file server and the usage of the same tag between the file server and the Spark code under test makes this workflow less than intuitive)

File: build/mvn
Comment (mccheah) [inline build/mvn:1]: Taken from https://github.com/apache/spark/blob/master/build/mvn

File: scripts/parse-arguments.sh
Comment (mccheah) [inline scripts/parse-arguments.sh]: A lot of these defaults are repeated again in `pom.xml` - the idea is that maintainers of the framework can run the scripts in isolation without using Maven. This can be done both with the top level `setup-integration-test-env.sh` script, or any of the sub-components - e.g. `clone-spark.sh`.

File: scripts/parse-arguments.sh
Comment (mccheah) [inline scripts/parse-arguments.sh]: This is less than ideal. It's more elegant to have here: `SPARK_TGZ=` and then to check in other places, `if [[ -z $SPARK_TGZ ]]`. The problem is that Maven's `exec` plugin has to pass in a fixed set of command line arguments. Maven will always have to pass in some `spark-tgz` argument. But we can't pass in empty, because we then end up with commands like `setup-integration-test-env --spark-tgz --deploy-mode cloud` which screws up the argument shifting.

An alternative we can employ in general is to allow these variables to be set by environment variables as well and to pass them in as environment variables from Maven. Ideas are welcome here.

File: scripts/write-docker-tag.sh
Comment (mccheah) [inline scripts/write-docker-tag.sh]: Less than ideal. The problem is that Maven wants to pass the image tag into `KubernetesSuite`, but Maven doesn't know it offhand until it's resolved by the test environment script. We use a UUID for the default image tag because there can be multiple builds concurrently running against the same Spark git hash. But the UUID can't be generated by Maven, meaning the script has to make it and pass that information along somehow, and a temporary file was the best bridge I could think of. Other more elegant ideas would be appreciated here.

**Comment by mccheah** [comment]
The diff is against a branch that is the outcome of merging master with #29 and #14.

File: e2e/e2e-prow.sh
Comment (foxish) [inline e2e/e2e-prow.sh:1]: This is specific to GKE testsuites, I think we still need this setup script, even if the runner goes away.

Comment (mccheah) [reply e2e/e2e-prow.sh:1]: Ah I didn't notice the apt-get calls. I think we want these to be assumed by the framework. Installing components on demand in a test framework is scary-ish - not to mention that installation probably depends on the user's OS, e.g. if the user is running from Mac OSX this won't work. We should have separate scripts that aren't invoked in the build lifecycle for these.

**Comment by foxish** [comment]
Agreed. It's separate and the only expected consumer is k8s test
infrastructure.

File: README.md
Comment (foxish) [inline README.md]: I liked before that it was using the git sha - it made it easy to correlate to the distribution that was just tested.

File: README.md
Comment (foxish) [inline README.md]: `gitRepo` and `gitBranch` may be better names for this.

File: build/mvn
Comment (foxish) [reply build/mvn:1]: Why do we need this? We'll need to ensure that it stays in sync with our repo? Or maybe we can get rid of it once we merge this upstream?

File: e2e/e2e-prow.sh
Comment (foxish) [reply e2e/e2e-prow.sh:1]: Yes, did not intend for these to be part of the testsuite. These are external "setup" steps and the entrypoint for the k8s CI system.

File: scripts/build-spark.sh
Comment (foxish) [inline scripts/build-spark.sh]: If we omit `--tgz`, we should just get the unpacked distribution under `dist/`.

File: scripts/parse-arguments.sh
Comment (foxish) [reply scripts/parse-arguments.sh]: Hmm.. it seems error prone that this needs to stay in sync with the args in pom.xml. Is there a better dispatch mechanism than to use this?

File: scripts/write-docker-tag.sh
Comment (foxish) [reply scripts/write-docker-tag.sh]: Concurrent runs against the same git-sha, I'm not quite sure. So, if a new PR is submitted, it would apply that commit on the base and then run the tests - which would make each one have a unique sha. If they are the same sha, then invariably, they are identical and the image could be reused.

File: src/test/scala/org/apache/spark/deploy/k8s/integrationtest/config.scala
Comment (foxish) [inline src/test/scala/org/apache/spark/deploy/k8s/integrationtest/config.scala]: These need to be cleaned up I think.

File: scripts/write-docker-tag.sh
Comment (mccheah) [reply scripts/write-docker-tag.sh]: Think there's two git hashes to consider and neither will work:

- If we're talking about the integration test repository git hash, multiple builds should certainly be able to run against the same commit. Multiple test runs on the same node testing two different commits of Spark but using the same integration test version comes to mind.
- If we're talking about the Spark repository git hash, the inverse rule applies here - multiple versions of the integration test running against the same Spark version. Reusing the images is difficult because we need to know when it's safe to GC the image, which is only if all tests holding that given tag are finished. Then we can run into race conditions, etc.

Seems a lot easier to bias towards isolation for correctness and optimize later.

File: build/mvn
Comment (mccheah) [reply build/mvn:1]: We get rid of this when we merge into upstream, but even aside from that, this has hardly been changing in upstream and it's theoretically completely isolated from what we will use to build Spark (since the shell forks the second Maven process it's completely independent)

File: README.md
Comment (mccheah) [reply README.md]: Discussed in https://github.com/apache-spark-on-k8s/spark-integration/pull/31#discussion_r161349039

File: scripts/build-spark.sh
Comment (mccheah) [reply scripts/build-spark.sh]: Makes it easier to reuse the distribution with the `spark.kubernetes.test.sparkTgz` option. Thus that configuration can pull double duty with custom Spark distributions and re-using dynamically built ones.

File: scripts/parse-arguments.sh
Comment (mccheah) [reply scripts/parse-arguments.sh]: The `pom.xml` can assume that these are the defaults and that they are correct, but I like having them explicitly enumerated in the `pom.xml` because the behavior can be known right from the pom. But I also like being able to run the script independently without any arguments. But all of this is just for convenience, so for the sake of developer sanity we can make one of these ideas less convenient.

Don't have a strong opinion on the best way to move forward.

**Comment by mccheah** [comment]
Changed the base to `master` and removed dependence on #29. Turns out we actually need this to run at all now - our tests are broken since the Docker image format change introduced in https://github.com/apache/spark/pull/20192

File: build/mvn
Comment (mccheah) [reply build/mvn:1]: We also have to add it here now because unlike before, now `mvn` is the entrypoint into the whole system, as it's only the substep of the build reactor that clones the Spark repository down. Aside from that, we also allow for Spark TGZs to be provided to remove the usage of the Spark source code entirely. So no matter what we're going to need our own Maven here.

File: scripts/setup-integration-test-env.sh
Comment (foxish) [inline scripts/setup-integration-test-env.sh]: Instead of passing all the args around, we should parge the args here, and then set the right env-vars/args and call the other scripts maybe? I think it's an antipattern for each script to be calling parse-args and seeing unrelated args.

File: scripts/parse-arguments.sh
Comment (ssuchter) [reply scripts/parse-arguments.sh]: You could use the maven exec plugin like this:

```
              <executable>/bin/sh</executable>
              <arguments>
                <argument>-c</argument>
                <argument>arbitrary length here ; another command</argument>
              </arguments>
```

File: README.md
Comment (ssuchter) [inline README.md]: I don't understand the motivation of why we should have the integration test code ever try to clone and build spark. Why not just always depend on a pre-built spark distribution? (Sorry if I missed the reason for this)

File: build/mvn
Comment (ssuchter) [reply build/mvn:1]: Instead of copying the content here, could you just have a much shorter script that just downloads from https://raw.githubusercontent.com/apache/spark/master/build/mvn and execs it? Something like build/mvn-getter whose content is:
```
#!/bin/sh
curl -s https://raw.githubusercontent.com/apache/spark/master/build/mvn | bash

File: scripts/build-spark.sh
Comment (ssuchter) [reply scripts/build-spark.sh]: I think for jenkins, we will want to use the pre-built distributions exclusively. This is because we don't want to keep the minikube locked during the building of the distribution. If the spark distribution builds "inside" of the spark-integration test, then it will be harder to express this partial locking. As it currently stands, one jenkins job creates the spark distribution, and another jenkins job does the integration test, making the locking easy.

File: scripts/write-docker-tag.sh
Comment (ssuchter) [reply scripts/write-docker-tag.sh]: Would it be simpler to just GC any images that are from this, but over a day old?

Comment (mccheah) [reply scripts/write-docker-tag.sh]: There's still a potential race with who gets to build the image with the given hash though. Two builds trying to use the same hash need to agree on who builds the image. I don't want to have to deal with that locking at all, so just use completely unique tags everywhere.

File: scripts/parse-arguments.sh
Comment (mccheah) [reply scripts/parse-arguments.sh]: The problem is that the arguments themselves have to be fixed. I can't sometimes pass in the `--spark-tgz` flag and sometimes not pass the `--spark-tgz` flag.

File: build/mvn
Comment (mccheah) [reply build/mvn:1]: Hm, I think that works. Will incorporate that.

File: README.md
Comment (mccheah) [reply README.md]: I've found it a lot easier locally to be able to run a single command that both fetches the Spark distribution and builds it and then the integration test uses that Spark distribution. We provide the optionality for the local development scenario but we can still provide the TGZ specifically in Jenkins via `spark.kubernetes.test.sparkTgz`.

File: scripts/parse-arguments.sh
Comment (ssuchter) [inline scripts/parse-arguments.sh]: Presumably if we don't try to build spark in here, we don't need BRANCH, SPARK_REPO, SPARK_REPO_LOCAL_DIR, UNPACKED_SPARK_TGZ (we might only have SPARK_TGZ, and it's always required), right?

File: scripts/parse-arguments.sh
Comment (ssuchter) [reply scripts/parse-arguments.sh]: This gets simpler with the idea that this code always takes --spark-tgz, I think.

File: scripts/setup-integration-test-env.sh
Comment (ssuchter) [inline scripts/setup-integration-test-env.sh]: I think if we have this always take a spark.tgz, then this branch isn't necessary

File: scripts/setup-integration-test-env.sh
Comment (ssuchter) [inline scripts/setup-integration-test-env.sh]: Maybe this would change to unpack-spark.sh?

File: scripts/setup-integration-test-env.sh
Comment (ssuchter) [inline scripts/setup-integration-test-env.sh]: If unpacking spark is something we always do, is there any reason to have this skip argument? Why else would one call this setup-integration-test-env.sh other than building the dockers?

File: scripts/write-docker-tag.sh
Comment (ssuchter) [reply scripts/write-docker-tag.sh]: Ok!

File: build/mvn
Comment (mccheah) [reply build/mvn:1]: Done

**Comment by mccheah** [comment]
@ssuchter @foxish approach has changed - the `dev/dev-run-integration-tests.sh` is for local running and is whats documented in `README.md`, which builds the Spark distribution and sends it to the Maven reactor. The maven reactor itself uses a separate script to build the Docker image, but the Maven reactor must be given the Spark distribution to run with.

**Comment by mccheah** [comment]
Reason we have to do this was discussed in sig-big-data - Jenkins will only use the Maven build without building the Spark distribution, so that the maven build doesn't lock the Minikube resource on building Spark, which takes a long time.

File: src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/GCE/GCETestBackend.scala
Comment (mccheah) [inline src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/GCE/GCETestBackend.scala:39]: Though the more I look at this abstraction, the more I realize it's not entirely necessary anymore. The biggest difference before was that using the MinikubeTestBackend triggered the Docker build, but now, we're building images regardless of the backend type. The Minikube version accepts the Minikube certificates, but those can be passed along as system properties. It's not something to change in this PR, but it's something that can be changed strictly because of the way the image build workflow functions now.

**Comment by mccheah** [comment]
@foxish @ssuchter the command in Jenkins is now incorrect. Seeing that all our other builds will fail now anyways, we might as well update the build command to reflect what this version of the tests expects. Then we can check if this build would be green.

The Jenkins instance ran this:

`/home/jenkins/tools/hudson.tasks.Maven_MavenInstallation/maven/bin/mvn -f integration-test/pom.xml clean integration-test -DzincPort=3096 -Dspark-distro-tgz=/home/jenkins/workspace/PR-dev-spark-integration/target/copied-artifacts-spark-distro/spark-2.4.0-SNAPSHOT-bin-20180112-55dbfbc.tgz`

Pretty sure it has to be changed to this now:

    <repo-root>/build/mvn clean integration-test -Dspark.kubernetes.test.sparkTgz=<spark tgz path> -DzincPort=3096

Main differences of course being, we now use our built-in Maven, and we can use the `pom.xml` file verbatim. But this PR also changes the property for the Spark distribution, mostly just for aesthetics.

File: dev/dev-run-integration-tests.sh
Comment (mccheah) [inline dev/dev-run-integration-tests.sh:89]: Would like to add support for arbitrary maven arguments, just couldn't quite get the shell scripting to work out yet.

File: build/mvn
Comment (ssuchter) [inline build/mvn]: Why not just pipe to bash?

File: dev/dev-run-integration-tests.sh
Comment (ssuchter) [inline dev/dev-run-integration-tests.sh:64]: Are you going to keep this clause, or delete it? I think it's pretty straightforward this way, so I think it'd be reasonable to keep it.

File: dev/dev-run-integration-tests.sh
Comment (ssuchter) [inline dev/dev-run-integration-tests.sh:21]: Remove unnecessary quotes

File: dev/dev-run-integration-tests.sh
Comment (ssuchter) [inline dev/dev-run-integration-tests.sh:22]: Remove unnecessary quotes

File: dev/dev-run-integration-tests.sh
Comment (ssuchter) [inline dev/dev-run-integration-tests.sh]: Remove unnecessary quotes

File: dev/dev-run-integration-tests.sh
Comment (ssuchter) [inline dev/dev-run-integration-tests.sh]: Remove unnecessary quotes

File: dev/dev-run-integration-tests.sh
Comment (ssuchter) [inline dev/dev-run-integration-tests.sh]: Remove unnecessary quotes

File: scripts/setup-integration-test-env.sh
Comment (ssuchter) [inline scripts/setup-integration-test-env.sh]: Remove unnecessary quotes

File: scripts/setup-integration-test-env.sh
Comment (ssuchter) [inline scripts/setup-integration-test-env.sh]: Remove unnecessary quotes

File: scripts/setup-integration-test-env.sh
Comment (ssuchter) [inline scripts/setup-integration-test-env.sh]: Remove unnecessary quotes

File: scripts/setup-integration-test-env.sh
Comment (ssuchter) [inline scripts/setup-integration-test-env.sh:63]: Remove unnecessary quotes

**Review by ssuchter — APPROVED** [review]
I think this looks nice. I mainly left some comments about removing some unneeded double quotes in shell expressions. :)

**Comment by ssuchter** [comment]
I should say to all of my comments about removing unnecessary quotes. I guess it'd also be reasonable to add quotes for all variable assignments (although they aren't necessary). I think consistency is important.

**Comment by kimoonkim** [comment]
@mccheah wrote
> @foxish @ssuchter the command in Jenkins is now incorrect. Seeing that all our other builds will fail now anyways, we might as well update the build command to reflect what this version of the tests expects. Then we can check if this build would be green.

This sounds good to me. Has anyone started fixing this? Otherwise, I'll take on that. Builds are all failing as predicted and we'd better fix it soon.

**Comment by kimoonkim** [comment]
rerun integration tests please

**Comment by kimoonkim** [comment]
rerun integration tests please

**Comment by foxish** [comment]
@kimoonkim, I don't think anyone has looked at the jenkins command yet. Would be good if we can fix it in time for this PR to merge and get things green again.

File: scripts/setup-integration-test-env.sh
Comment (foxish) [inline scripts/setup-integration-test-env.sh]: This would already be done by the `-m` option.

**Comment by kimoonkim** [comment]
Cool. Working on it.

File: dev/dev-run-integration-tests.sh
Comment (kimoonkim) [inline dev/dev-run-integration-tests.sh]: Should drop the backslash here.

File: dev/dev-run-integration-tests.sh
Comment (kimoonkim) [inline dev/dev-run-integration-tests.sh]: Should drop the backslash.

**Review by kimoonkim — CHANGES_REQUESTED** [review]

**Comment by kimoonkim** [comment]
rerun integration tests please

**Comment by kimoonkim** [comment]
rerun integration tests please

**Comment by kimoonkim** [comment]
rerun integration tests please

**Comment by kimoonkim** [comment]
rerun integration tests please

**Comment by kimoonkim** [comment]
rerun integration tests please

**Comment by kimoonkim** [comment]
rerun integration tests please

**Comment by kimoonkim** [comment]
rerun integration tests please

**Comment by kimoonkim** [comment]
rerun integration tests please

**Comment by kimoonkim** [comment]
Ok. I made Jenkins use dev-run-integration-tests.sh, which get us the following error:
> dev/dev-run-integration-tests.sh --spark-tgz /home/jenkins/workspace/PR-dev-spark-integration/target/copied-artifacts-spark-distro/spark-2.4.0-SNAPSHOT-bin-20180113-fc6fe8a.tgz
dev/dev-run-integration-tests.sh: line 101: syntax error: unexpected end of file

I think it's due to the bug pointed out [here](https://github.com/apache-spark-on-k8s/spark-integration/pull/31#pullrequestreview-88665624)

**Comment by kimoonkim** [comment]
rerun integration tests please

**Comment by kimoonkim** [comment]
rerun integration tests please

**Comment by kimoonkim** [comment]
rerun integration tests please

**Comment by mccheah** [comment]
@kimoonkim the `dev/dev-run-integration-tests.sh` script is for development setups, though I took a look at the command and that command should theoretically work in Jenkins as well.

Couple of questions:

- Minikube appears to be starting in the output, shouldn't it already be running?
- Minikube is stopping at the end. Shouldn't we be keeping it running?

File: build/mvn
Comment (mccheah) [reply build/mvn]: Do the arguments sent to this script follow the pipe?

Comment (mccheah) [reply build/mvn]: We also don't want to download the mvn script if we already have it - so the script ends up writing simpler with that fix. Addressed in latest patch.

**Comment by mccheah** [comment]
@ssuchter let's just have quotes everywhere.

**Comment by mccheah** [comment]
@kimoonkim the `minikube` binary needs to be on the `PATH`.

**Comment by mccheah** [comment]
Sorry, forgot to mention what the specific command should be in Jenkins without using the dev script: `build/mvn integration-test -Dspark.kubernetes.test.sparkTgz=<tgz-path>`.

**Comment by mccheah** [comment]
rerun integration tests please

**Comment by mccheah** [comment]
@kimoonkim integration tests aren't running when I comment, can you trigger the build?

File: dev/dev-run-integration-tests.sh
Comment (ssuchter) [reply dev/dev-run-integration-tests.sh]: Duh, it's just the last line (line 92) where it should be dropped, I got it now.

**Comment by kimoonkim** [comment]
> @kimoonkim integration tests aren't running when I comment, can you trigger the build?

I think this build [98](http://spark-k8s-jenkins.pepperdata.org:8080/job/PR-dev-spark-integration/98/) actually ran upon your comment, given the timing of build start. I believe there is a quiet period of dozen secs. So maybe it did not trigger right away.

**Review by ssuchter — COMMENTED** [review]
.

**Comment by kimoonkim** [comment]
> Minikube appears to be starting in the output, shouldn't it already be running?
Minikube is stopping at the end. Shouldn't we be keeping it running?

Right. It gets stopped and deleted at the end. Pepperdata Jenkins cluster has only one node and it still has jobs running old integration tests from our fork. If we keep minikube running, that will interfere with the old integration tests. Stopping minikube is the best compromise for this setup, IMO.

**Comment by mccheah** [comment]
Hm, but the error is this:

```
  java.io.IOException: Cannot run program "minikube": error=2, No such file or directory
  at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
```

But the tests work locally for me. I wonder if Java's `ProcessBuilder` doesn't use the `PATH` to find locations of executables.

**Comment by kimoonkim** [comment]
> Sorry, forgot to mention what the specific command should be in Jenkins without using the dev script: build/mvn integration-test -Dspark.kubernetes.test.sparkTgz=<tgz-path>.

Jenkins can switch to this command, but what is the advised way of building docker images now that the scala code does not build them?

**Comment by mccheah** [comment]
@kimoonkim this PR puts us in an in-between state where Maven is building the Docker images but we expect the external Jenkins job to build the Spark distribution. If we build using Maven, the reactor will still build the images for us.

**Comment by kimoonkim** [comment]
> Hm, but the error is this:
>>
>>  java.io.IOException: Cannot run program "minikube": error=2, No such file or directory
  at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
> But the tests work locally for me. I wonder if Java's ProcessBuilder doesn't use the PATH to find locations of executables.

I see I'll try adding the minikube location to the PATH used by maven.

**Comment by kimoonkim** [comment]
rerun integration tests please

**Comment by kimoonkim** [comment]
Now, the error from build [101](http://spark-k8s-jenkins.pepperdata.org:8080/job/PR-dev-spark-integration/101/consoleFull) is this:
> /home/jenkins/workspace/PR-dev-spark-integration/target/spark-dist-unpacked/bin/docker-image-tool.sh: line 66: docker: command not found

@ssuchter I believe you installed minikube on Pepperdata Jenkins. Did we install docker?

**Comment by ssuchter** [comment]
We're working on that same issue (missing docker) in slack sig-big-data

**Comment by mccheah** [comment]
rerun integration tests please

**Comment by mccheah** [comment]
rerun integration tests please

File: dev/dev-run-integration-tests.sh
Comment (ssuchter) [reply dev/dev-run-integration-tests.sh]: I think this is addressed.

**Pull request merged**
